src/codec.cairo

```diff
@@ -58,26 +100,60 @@ pub impl EncodeTxIn of Encode<TxIn> {
         (*self.script).encode_to(ref dest);
         self.sequence.encode_to(ref dest);
     }
+
+    fn weight_to(self: @TxIn, ref weight: usize, factor: usize) {
+        self.previous_output.weight_to(ref weight, factor);
+        (*self.script).weight_to(ref weight, factor);
+        self.sequence.weight_to(ref weight, factor);
+    }
 }
 
 pub impl EncodeTxOut of Encode<TxOut> {
     fn encode_to(self: @TxOut, ref dest: ByteArray) {
         self.value.encode_to(ref dest);
         (*self.pk_script).encode_to(ref dest);
     }
+
+    fn weight_to(self: @TxOut, ref weight: usize, factor: usize) {
+        self.value.weight_to(ref weight, factor);
+        (*self.pk_script).weight_to(ref weight, factor);
+    }
 }
 
 pub impl EncodeOutpoint of Encode<OutPoint> {
     fn encode_to(self: @OutPoint, ref dest: ByteArray) {
         self.txid.encode_to(ref dest);
         self.vout.encode_to(ref dest);
     }
+
+    fn weight_to(self: @OutPoint, ref weight: usize, factor: usize) {
+        self.txid.weight_to(ref weight, factor);
+        self.vout.weight_to(ref weight, factor);
+    }
 }
 
 pub impl EncodeTransaction of Encode<Transaction> {
     fn encode_to(self: @Transaction, ref dest: ByteArray) {
         self.encode_transaction_to(ref dest, false);
     }
+
+    fn weight_to(self: @Transaction, ref weight: usize, factor: usize) {
+        self.version.weight_to(ref weight, NON_WITNESS_FACTOR);
+        if (*self.is_segwit) {
+            weight += 1; // marker
+            weight += 1; // flag
+        }
+
+        self.inputs.weight_to(ref weight, NON_WITNESS_FACTOR);
+        self.outputs.weight_to(ref weight, NON_WITNESS_FACTOR);
+        if (*self.is_segwit) {
+            for txin in *self.inputs {
+                txin.witness.weight_to(ref weight, WITNESS_FACTOR);
```

Comment:
Witness factor is used only here, it makes sense not to introduce it as an argument to the weight